### PR TITLE
MEN-4427: Generate configuration artifacts on-the-fly on /deployments/next requests

### PR DESCRIPTION
**Pull request by alfrunes**
This PR implements a new endpoint consuming signed expiring links generated by the /deployments/next API handler if the deployment is a configuration type deployment.

NOTE: This PR is still work in progress as there's missing some unit test coverage.

### CODE REVIEW

**Review by kjaskiewiczz — APPROVED** [review]
great job!
the task was very complex IMO and you managed to finish it very quickly!
I can't see issues with this implementation, but as I mentioned, it's complex so it would be great if one more person could take a look.

File: config/config.go [hunk 1]
```diff
@@ -73,6 +73,33 @@ const (
...
+	// downloadable URLs. This option is currently ignored.
+	SettingPresignAlgorithm        = "presign.algorithm"
+	SettingPresignAlgorithmDefault = "HMAC256"
+
+	// SettingPresignSecret sets the secret for generating signed url.
+	// For HMAC type of algorithms the value must be a base64 encoded
+	// secret. For public key signatures, the value must be a path to
+	// the private key (not yet supported).
+	SettingPresignSecret        = "presign.secret"
+	SettingPresignSecretDefault = "qF4/MZzQmHTJ+nrlr26b3g=="
```
Comment (tranchitella) [inline config/config.go:87]: I don't think it's a good idea to provide a default here, as this can lead to security issues.
Can we find a better way to deal with it? A random value at start-up is good enough, IMHO.

Comment (alfrunes) [reply config/config.go]: Good call!
Actually I discovered that I incorrectly configured the defaults below as well.
I pushed new commit generating making the secret default to a random base64 string of 32 bytes if not configured explicitly.

**Review by tranchitella — APPROVED** [review]

File: api/http/api_deployments.go [hunk 2]
```diff
@@ -236,6 +265,80 @@ func (d *DeploymentsApiHandlers) DownloadLink(w rest.ResponseWriter, r *rest.Req
...
+	if d.config.PresignSecret == nil {
+		rest.NotFound(w, r)
+		return
+	}
+
+	var (
+		tenantID     string
+		deviceID     = r.PathParam("device_id")
+		deviceType   = r.PathParam("device_type")
+		deploymentID = r.PathParam("deployment_id")
```
Comment (merlin-northern) [inline api/http/api_deployments.go:315]: I would really love to see those as consts.

File: api/http/api_deployments.go [hunk 3]
```diff
@@ -236,6 +265,80 @@ func (d *DeploymentsApiHandlers) DownloadLink(w rest.ResponseWriter, r *rest.Req
...
+		l            = log.FromContext(r.Context())
+		q            = r.URL.Query()
+		err          error
+	)
+	tenantID = q.Get("tenant_id")
```
Comment (merlin-northern) [inline api/http/api_deployments.go:320]: I would really love to see that as a const/var

File: api/http/api_deployments.go [hunk 4]
```diff
@@ -994,6 +1005,26 @@ func (d *DeploymentsApiHandlers) GetDeploymentForDevice(w rest.ResponseWriter, r
 	if deployment == nil {
 		d.view.RenderNoUpdateForDevice(w)
 		return
+	} else if deployment.Type == model.DeploymentTypeConfiguration {
+		// Generate pre-signed URL
+		hostName := r.Header.Get("X-Forwarded-Host")
```
Comment (merlin-northern) [inline api/http/api_deployments.go]: a const ehre, perhaps?

File: api/http/routing.go [hunk 5]
```diff
@@ -102,9 +105,15 @@ func NewRouter(ctx context.Context, c config.Reader,
 
 	app := app.NewDeployments(mongoStorage, fileStorage, app.ArtifactContentType)
 
-	apiConf := NewConfig()
-	if key, err := base64.StdEncoding.DecodeString(
-		c.GetString(dconfig.SettingPresignSecretBase64),
+	// Encode base64 secret in either std or URL encoding ignoring padding.
+	base64Repl := strings.NewReplacer("-", "+", "_", "/", "=", "")
```
Comment (merlin-northern) [inline api/http/routing.go:111]: Maybe it is late, but I do not understand this replacing, why is it so?

File: api/http/routing.go [hunk 6]
```diff
@@ -222,3 +231,12 @@ func ReleasesRoutes(controller *DeploymentsApiHandlers) []*rest.Route {
 		rest.Get(ApiUrlManagementReleases, controller.GetReleases),
 	}
 }
+
+func FMTConfigURL(hostname, deploymentID, deviceType, deviceID string) string {
+	repl := strings.NewReplacer(
+		":deployment_id", deploymentID,
+		":device_type", deviceType,
+		":device_id", deviceID,
```
Comment (merlin-northern) [inline api/http/routing.go:246]: on more reason to have those in consts.

File: api/http/routing.go [hunk 7]
```diff
@@ -222,3 +231,12 @@ func ReleasesRoutes(controller *DeploymentsApiHandlers) []*rest.Route {
...
+	)
+	return fmt.Sprintf("https://%s%s", hostname, repl.Replace(ApiUrlDevicesDownloadConfig))
```
Comment (merlin-northern) [inline api/http/routing.go]: Isn't Sprintf and over kill here? `"https://"+hostname+repl.Replace(ApiUrlDevicesDownloadConfig)` perhaps?

File: api/http/api_deployments.go [hunk 8]
```diff
@@ -85,13 +85,23 @@ var (
...
 type Config struct {
+	// URL signing parameters:
+
+	// PresignSecret holds the secret value used by the signature algorithm.
 	PresignSecret []byte
+	// PresignExpire duration until the link expires.
 	PresignExpire time.Duration
+	// PresignHostname fallback value for the signed url hostname.
+	PresignHostname string
+	// PresignScheme URL scheme to used for generating signed URLs.
```
Comment (merlin-northern) [inline api/http/api_deployments.go:96]: nitpick: "to used" -> "to _be_ used" ot "to _use_"?

File: api/http/api_deployments.go [hunk 9]
```diff
@@ -85,13 +85,23 @@ var (
...
+	PresignScheme string
 }
 
 func NewConfig() *Config {
 	return &Config{
-		PresignExpire: time.Second * 900,
+		PresignExpire:   time.Second * 900,
```
Comment (merlin-northern) [inline api/http/api_deployments.go:102]: I know, not part of the commit, but as long as you are here, maybe some explanation on the value of 900 and a const?

File: api/http/api_deployments.go [hunk 10]
```diff
@@ -1007,10 +1033,17 @@ func (d *DeploymentsApiHandlers) GetDeploymentForDevice(w rest.ResponseWriter, r
 		return
 	} else if deployment.Type == model.DeploymentTypeConfiguration {
 		// Generate pre-signed URL
-		hostName := r.Header.Get("X-Forwarded-Host")
+		var hostName string
+		if hostName = r.Header.Get("X-Forwarded-Host"); hostName == "" {
```
Comment (merlin-northern) [inline api/http/api_deployments.go:1051]: already used second time, maybe a const? (or this commit changes the previous)

File: api/http/api_deployments.go [hunk 11]
```diff
@@ -1007,10 +1033,17 @@ func (d *DeploymentsApiHandlers) GetDeploymentForDevice(w rest.ResponseWriter, r
...
+			hostName = d.config.PresignHostname
+		}
 		req, _ := http.NewRequest(
 			http.MethodGet,
-			FMTConfigURL(hostName, deployment.ID, installed.DeviceType, idata.Subject),
+			FMTConfigURL(
+				d.config.PresignScheme, hostName,
+				deployment.ID, installed.DeviceType,
+				idata.Subject,
+			),
```
Comment (merlin-northern) [inline api/http/api_deployments.go:1060]: and the reason behind this formatting?

File: api/http/routing.go [hunk 12]
```diff
@@ -232,11 +238,11 @@ func ReleasesRoutes(controller *DeploymentsApiHandlers) []*rest.Route {
...
 
-func FMTConfigURL(hostname, deploymentID, deviceType, deviceID string) string {
+func FMTConfigURL(scheme, hostname, deploymentID, deviceType, deviceID string) string {
 	repl := strings.NewReplacer(
 		":deployment_id", deploymentID,
 		":device_type", deviceType,
 		":device_id", deviceID,
 	)
-	return fmt.Sprintf("https://%s%s", hostname, repl.Replace(ApiUrlDevicesDownloadConfig))
+	return fmt.Sprintf("%s://%s%s", scheme, hostname, repl.Replace(ApiUrlDevicesDownloadConfig))
```
Comment (merlin-northern) [inline api/http/routing.go:248]: now I am sure the commits change the previous commits. the same comment: how about simple concat?

File: config/config.go [hunk 13]
```diff
@@ -74,11 +74,32 @@ const (
...
+
+	// SettingPresignExpireSeconds sets the amount of seconds it takes for
+	// the signed URL to expire.
+	SettingPresignExpireSeconds        = "presign.expire_seconds"
 	SettingPresignExpireSecondsDefault = 900
```
Comment (merlin-northern) [inline config/config.go:92]: I am quite sure I saw 900 in the code as a magic number.

File: config/config.go [hunk 14]
```diff
@@ -144,7 +165,10 @@ var (
 		{Key: SettingsAwsTagArtifact, Value: SettingsAwsTagArtifactDefault},
 		{Key: SettingInventoryAddr, Value: SettingInventoryAddrDefault},
 		{Key: SettingInventoryTimeout, Value: SettingInventoryTimeoutDefault},
-		{Key: SettingPresignSecretBase64, Value: SettingPresignSecretBase64},
+		{Key: SettingPresignAlgorithm, Value: SettingPresignAlgorithmDefault},
+		{Key: SettingPresignSecret, Value: SettingPresignSecret},
```
Comment (merlin-northern) [inline config/config.go]: shouldn't the second one be `SettingPresignSecretDefault` or something similar?

File: main.go [hunk 15]
```diff
@@ -90,6 +91,16 @@ func doMain(args []string) {
 		config.Config.SetEnvPrefix("DEPLOYMENTS")
 		config.Config.AutomaticEnv()
 		config.Config.SetEnvKeyReplacer(strings.NewReplacer(".", "_", "-", "_"))
+		if config.Config.Get(dconfig.SettingPresignSecret) ==
+			dconfig.SettingPresignSecretDefault {
+			l.Warn("Using the default signature secret; THIS IS A POTENTIAL SECURITY ISSUE!")
+			l.Warnf(
+				"Please reconfigure the secret in '%s' or "+
+					"override with environment variable "+
+					"DEPLOYMENTS_PRESIGN_SECRET_BASE64",
```
Comment (merlin-northern) [inline main.go]: that formatting seems strange just a bit?

File: api/http/api_deployments_test.go [hunk 16]
```diff
@@ -17,20 +17,24 @@ package http
...
 	"github.com/mendersoftware/deployments/app"
 	mapp "github.com/mendersoftware/deployments/app/mocks"
 	"github.com/mendersoftware/deployments/model"
 	"github.com/mendersoftware/deployments/utils/restutil/view"
 	h "github.com/mendersoftware/deployments/utils/testing"
 	"github.com/mendersoftware/go-lib-micro/requestid"
 	"github.com/mendersoftware/go-lib-micro/rest_utils"
+	"github.com/pkg/errors"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/mock"
```
Comment (merlin-northern) [inline api/http/api_deployments_test.go]: don't we group mendersoftware imports separately?

File: app/images_test.go [hunk 17]
```diff
@@ -534,3 +538,136 @@ func TestGenerateImageSuccessfulWithTenant(t *testing.T) {
...
+			uuid.NameSpaceOID,
+			[]byte("deployment"),
+		).String(),
+		Deployment: &model.Deployment{
+			Id: uuid.NewSHA1(
+				uuid.NameSpaceOID,
+				[]byte("deployment"),
+			).String(),
+			Type:          model.DeploymentTypeConfiguration,
+			Configuration: []byte("{\"foo\":\"bar\"}"),
```
Comment (merlin-northern) [inline app/images_test.go:568]: at the moment we are only supporting string values, I wonder how much of it should be tested here? perhaps a test that shows that we do not even car if it is json?

File: app/app.go [hunk 18]
```diff
@@ -45,8 +45,8 @@ import (
 const (
 	ArtifactContentType              = "application/vnd.mender-artifact"
 	ArtifactConfigureType            = "mender-configure"
-	ArtifactConfigureProvides        = "module-image.mender-configure.version"
-	ArtifactConfigureProvidesCleared = "module-image.mender-configure.*"
+	ArtifactConfigureProvides        = "data-partition.mender-configure.version"
+	ArtifactConfigureProvidesCleared = "data-partition.mender-configure.*"
```
Comment (merlin-northern) [inline app/app.go:49]: I would really enjoy some comments here, where that comes from, etc.

File: config/config.go (hunk 1 above)
Comment (mchalski) [reply config/config.go]: +1, that was my comment from writing acceptance tests - the default wasn't picked up, and it was 'good' imo.

File: api/http/api_deployments.go [hunk 19]
```diff
@@ -891,50 +1045,38 @@ func (d *DeploymentsApiHandlers) GetDeploymentForDevice(w rest.ResponseWriter, r
 	if deployment == nil {
 		d.view.RenderNoUpdateForDevice(w)
 		return
+	} else if deployment.Type == model.DeploymentTypeConfiguration {
+		// Generate pre-signed URL
+		var hostName string
+		if hostName = r.Header.Get("X-Forwarded-Host"); hostName == "" {
```
Comment (mchalski) [inline api/http/api_deployments.go:1051]: very good idea on having the presign host, and even the presign scheme. you could say we both honor the X-Forwarded header, and have a configuration fallback, right?

I think I would revert the logic here though. what if we promote the config setting:
- if the setting is there - use that
- if the setting is not there - use X-Forwarded-Host
- otherwise, fail

 the reason is I'm having second thoughts. I know I cofirmed that we can trust X-Forwarded-Host, because I'm pretty sure we can trust traefik to set it. I don't know however if in a live aws setup, with it's load balancer, we will get the original DNS name of the gateway here.
 
this will be researched and POCed still, so just a thought.

**Review by mchalski — APPROVED** [review]
great job.
I'm ready to approve, just left 1 discussion point

File: api/http/routing.go (hunk 5 above)
Comment (alfrunes) [reply api/http/routing.go:111]: It converts `base64.URLEncoding` to `base64.RawStdEncoding`. This way we can decode all 4 variants of base64 encoding (with or without padding).

File: api/http/api_deployments.go (hunk 19 above)
Comment (alfrunes) [reply api/http/api_deployments.go:1051]: Fair point, I'll change the logic to prefer the configuration setting and blank out the default config value.

File: api/http/api_deployments_test.go (hunk 16 above)
Comment (alfrunes) [reply api/http/api_deployments_test.go]: I'm not sure why it wasn't formatted like that here...

File: main.go (hunk 15 above)
Comment (alfrunes) [reply main.go]: Suggestions?

File: api/http/routing.go (hunk 12 above)
Comment (alfrunes) [reply api/http/routing.go:248]: Sure

File: api/http/api_deployments.go (hunk 9 above)
Comment (alfrunes) [reply api/http/api_deployments.go:102]: It's just the default from the config file.

Comment (alfrunes) [reply api/http/api_deployments.go:102]: LOL, I had created a constant without even using it :see_no_evil:

File: api/http/api_deployments.go (hunk 11 above)
Comment (alfrunes) [reply api/http/api_deployments.go:1060]: Suggestions?

**Comment by alfrunes** [comment]
I've addressed your comments @mchalski and @merlin-northern and rebased on top of current master as there was a small merge conflict.

**Comment by merlin-northern** [comment]
> I've addressed your comments @mchalski and @merlin-northern and rebased on top of current master as there was a small merge conflict.

right! thanks!

**Review by merlin-northern — APPROVED** [review]
